### PR TITLE
Raise an error when the cookiecutter context file is not generated

### PR DESCRIPTION
**Pull request by KostyaEsmukov**
I have spent some time debugging with @erastov why the cookiecutter context is not getting read for a new template. Turned out the context file generation was missing in the template (see https://scaraplate.readthedocs.io/en/latest/template.html#scaraplate.cookiecutter.ScaraplateConf).

This PR adds a check which ensures that the cookiecutter context file exists in the generated temporary project and raises a meaningful error otherwise. This would help in finding out the cause much quicker.

### CODE REVIEW

**Comment by erastov** [comment]
May be write a similar thing for setup.cfg? And yet, can you say what can be the use cases of scaraplate.conf, what can be added there?

File: src/scaraplate/rollup.py [hunk 1]
```diff
@@ -158,6 +160,29 @@ def get_target_project_cookiecutter_context(
         return dict(context)
 
 
+def ensure_cookiecutter_context_exists(
+    target_path: Path, scaraplate_yaml: ScaraplateYaml
+) -> None:
+    cookiecutter_context = scaraplate_yaml.cookiecutter_context_type(target_path)
+    basename = Path(f"{cookiecutter_context}").relative_to(target_path)
```
Comment (kxepal) [inline src/scaraplate/rollup.py]: Quite abusive way to do `str(thing)`.

File: src/scaraplate/rollup.py [hunk 2]
```diff
@@ -14,7 +14,7 @@
 from .template import TemplateMeta, get_template_meta_from_git
 
 
-__all__ = ("rollup",)
+__all__ = ("rollup", "InvalidScaraplateTemplateError")
```
Comment (kxepal) [inline src/scaraplate/rollup.py]: This is the only exception we want to public? I think we don't any.

**Comment by KostyaEsmukov** [comment]
> May be write a similar thing for setup.cfg? 

`cookiecutter_context` is an instance of [CookieCutterContext](https://scaraplate.readthedocs.io/en/latest/template.html#scaraplate.cookiecutter.CookieCutterContext), which currently has two implementations: [SetupCfg](https://scaraplate.readthedocs.io/en/latest/template.html#scaraplate.cookiecutter.SetupCfg) for setup.cfg and [ScaraplateConf](https://scaraplate.readthedocs.io/en/latest/template.html#scaraplate.cookiecutter.ScaraplateConf)  for .scaraplate.conf. The one which should be used is set in [scaraplate.yaml](https://scaraplate.readthedocs.io/en/latest/template.html#module-scaraplate.cookiecutter), default is ScaraplateConf.

All this means that this check would work for both setup.cfg and .scaraplate.conf, depending on which implementation is active.

> And yet, can you say what can be the use cases of scaraplate.conf, what can be added there?

Could you elaborate?

File: src/scaraplate/rollup.py (hunk 2 above)
Comment (KostyaEsmukov) [reply src/scaraplate/rollup.py]: This is the only custom exception as for now, so there's nothing else to publish.

File: src/scaraplate/rollup.py (hunk 1 above)
Comment (KostyaEsmukov) [reply src/scaraplate/rollup.py]: This is 2019.

File: src/scaraplate/rollup.py (hunk 2 above)
Comment (kxepal) [reply src/scaraplate/rollup.py]: `__all__` means public. Public means to preserve backward compatibility. There are some holywars about what to count as public on python ML, but I would like to reduce a number of those occurances for us if we don't mean such.

**Comment by erastov** [comment]
@KostyaEsmukov After that I wrote a comment, carefully read the code, and the questions disappeared, thanks!

File: src/scaraplate/rollup.py (hunk 1 above)
Comment (kxepal) [reply src/scaraplate/rollup.py]: Please take me back to my shiny 2008.

File: src/scaraplate/rollup.py (hunk 2 above)
Comment (KostyaEsmukov) [reply src/scaraplate/rollup.py]: > __all__ means public. Public means to preserve backward compatibility. 

Correct. And I indeed want to expose this exception as a part of the public API (and keep it in the future).

The goal is to allow scaraplate to be used in Python, and for that one might want to import this exception to process it differently.

File: src/scaraplate/rollup.py (hunk 1 above)
Comment (kxepal) [reply src/scaraplate/rollup.py]: The only qustion remains if that we can hive an error by such cast without worry of mypy. Technically, we shouldn't cast here anything anyway.

File: src/scaraplate/rollup.py (hunk 2 above)
Comment (kxepal) [reply src/scaraplate/rollup.py]: Good call. Than it's ok.

Comment (kxepal) [reply src/scaraplate/rollup.py]: But in this case exception shouldn't be declarared at the end of file.

File: src/scaraplate/rollup.py [hunk 3]
```diff
@@ -158,6 +160,29 @@ def get_target_project_cookiecutter_context(
         return dict(context)
 
 
+def ensure_cookiecutter_context_exists(
```
Comment (erastov) [inline src/scaraplate/rollup.py]: It seems to me, or is this logic a little duplicate of `get_target_project_cookiecutter_context`?

File: src/scaraplate/rollup.py (hunk 1 above)
Comment (KostyaEsmukov) [reply src/scaraplate/rollup.py]: `cookiecutter_context` has `__str__` which returns a full absolute path to the file. The goal of this line is to get a path relative to the temp directory.

Technically we should be fine with this, as `__str__` is considered to be a part of the public API, so it shouldn't change.

Also the result of this operation is checked in tests (which ensure that the exception message contains just `.scaraplate.conf`).

File: src/scaraplate/rollup.py (hunk 2 above)
Comment (KostyaEsmukov) [reply src/scaraplate/rollup.py]: > But in this case exception shouldn't be declarared at the end of file.

Fair enough. Moved to the beginning of the file and added a simple docstring.

File: src/scaraplate/rollup.py (hunk 1 above)
Comment (kxepal) [reply src/scaraplate/rollup.py]: Hm...can we avoid this magic by using somethings like `cookiecutter_context.path` which explicitly guarantee that we'll receive a path, not a some string representation?

Comment (KostyaEsmukov) [reply src/scaraplate/rollup.py]: Pushed a different solution: https://github.com/rambler-digital-solutions/scaraplate/compare/73be435f3e95958d7a9b820d3be9f4a3feecd8a3..3171161a71fd6385ea67e53533a9a638884b4bb3

What I don't like about the previous one is that a custom implementation of CookieCutterContext might not override `__str__` or might not use paths at all, so we shouldn't assume that `__str__` is always a path.

File: src/scaraplate/rollup.py (hunk 3 above)
Comment (KostyaEsmukov) [reply src/scaraplate/rollup.py]: Not exactly. Both these functions just wrap `cookiecutter_context.read()` and provide a meaningful output depending on the result. And the output in both cases is different.

For example, `get_target_project_cookiecutter_context` prints the context, but we don't need that in `ensure_cookiecutter_context_exists`.

File: src/scaraplate/rollup.py (hunk 1 above)
Comment (kxepal) [reply src/scaraplate/rollup.py]: LGFM, thank you! :beer:

File: src/scaraplate/rollup.py (hunk 3 above)
Comment (erastov) [reply src/scaraplate/rollup.py]: ok

**Pull request merged**